daemon/cluster/listen_addr.go

```diff
@@ -0,0 +1,159 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var interfaceAddr4, interfaceAddr6 net.IP
+
+	for _, addr := range addrs {
+		ipAddr, ok := addr.(*net.IPNet)
+
+		if ok {
+			if ipAddr.IP.To4() != nil {
+				// IPv4
+				if interfaceAddr4 != nil {
+					return nil, fmt.Errorf("interface %s has more than one IPv4 address", specifiedInterface)
+				}
+				interfaceAddr4 = ipAddr.IP
+			} else {
+				// IPv6
+				if interfaceAddr6 != nil {
+					return nil, fmt.Errorf("interface %s has more than one IPv6 address", specifiedInterface)
+				}
+				interfaceAddr6 = ipAddr.IP
+			}
+		}
+	}
+
+	if interfaceAddr4 == nil && interfaceAddr6 == nil {
+		return nil, fmt.Errorf("interface %s has no usable IPv4 or IPv6 address", specifiedInterface)
+	}
+
+	// In the case that there's exactly one IPv4 address
+	// and exactly one IPv6 address, favor IPv4 over IPv6.
+	if interfaceAddr4 != nil {
+		return interfaceAddr4, nil
+	}
+	return interfaceAddr6, nil
+}
+
+var errMultipleIPs = errors.New("could not choose a listening IP address since this system has multiple addresses - specify one with --listen-addr or --listen-interface")
+var errNoIP = errors.New("could not find the system's IP address - specify one with --listen-addr or --listen-interface")
+
+func resolveSystemAddr() (net.IP, error) {
+	// Use the system's only IP address, or fail if there are
+	// multiple addresses to choose from.
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+
+	var systemAddr net.IP
+
+	for _, intf := range interfaces {
+		addrs, err := intf.Addrs()
+		if err != nil {
+			continue
+		}
+
+		var interfaceAddr4, interfaceAddr6 net.IP
+
+		for _, addr := range addrs {
+			ipAddr, ok := addr.(*net.IPNet)
+
+			// Skip loopback and link-local addresses
+			if ok && isUsableSystemIP(ipAddr.IP) {
+				if ipAddr.IP.To4() != nil {
+					// IPv4
+					if interfaceAddr4 != nil {
+						return nil, errMultipleIPs
+					}
+					interfaceAddr4 = ipAddr.IP
+				} else {
+					// IPv6
+					if interfaceAddr6 != nil {
+						return nil, errMultipleIPs
+					}
+					interfaceAddr6 = ipAddr.IP
+				}
+			}
+		}
+
+		// In the case that this interface has exactly one IPv4 address
+		// and exactly one IPv6 address, favor IPv4 over IPv6.
+		if interfaceAddr4 != nil {
+			if systemAddr != nil {
+				return nil, errMultipleIPs
+			}
+			systemAddr = interfaceAddr4
+		} else if interfaceAddr6 != nil {
+			if systemAddr != nil {
+				return nil, errMultipleIPs
+			}
+			systemAddr = interfaceAddr6
+		}
+	}
+
+	if systemAddr == nil {
+		return nil, errNoIP
+	}
+
+	return systemAddr, nil
+}
+
+var ipv4loopback = net.IPv4(127, 0, 0, 1)
+
+// isUsableSystemIP is similar to (IP).IsGlobalUnicast, but it checks specifically
+// for 127.0.0.1 instead of 127.*, because there are cases where a 127.x.x.x
+// address is the correct one to use.
+func isUsableSystemIP(ip net.IP) bool {
+	return !ip.Equal(net.IPv4bcast) &&
+		!ip.IsUnspecified() &&
+		!ip.Equal(ipv4loopback) &&
```

Comment:
As I mentioned in https://github.com/docker/docker/pull/24237/files/8368703fca7cfcfb9088e274cb1fbc300ac86aec#r69358757, I'm explicitly avoding this function because it breaks Docker-in-Docker use cases.